From a4532418fc8435349b54b7d94ef28020db6e6c8a Mon Sep 17 00:00:00 2001
From: Yu Xia <xiax.yu@intel.com>
Date: Fri, 29 Apr 2016 16:31:14 +0000
Subject: [PATCH 1/2] Input: touch: silead: Solve touch reading to detect 10
 fingers.

The root cause of this issue: i2c_smbus_read_i2c_block_data
function can read up to I2C_SMBUS_BLOCK_MAX (It was defined
 32 bytes as specified in SMBus standard) bytes at a time.
But we use i2c_smbus_read_i2c_block_data function to read
0x80 (counter about finger numbers) and 10 points coordinates.

In fact, A total of 44 bytes are required to read.
So the extra 12 bytes are not normally read, and the final 3
 points of the coordinates can not be detected.

The solution of this issue:
We should use i2c_smbus_read_i2c_block_data function to read
 twice times. Read for the first time 12 bytes, and finally
read the 32 bytes.

Change-Id: I293b2dc1edd1b1c708365cf85ef54adbd8746302
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-30193
Signed-off-by: Yu Xia <xiax.yu@intel.com>
Signed-off-by: Ulises Mendez Martinez <ulises.mendez.martinez@intel.com>
Reviewed-on: https://android.intel.com/#/c/501683/
Reviewed-on: https://android.intel.com:443/501683
---
 drivers/input/touchscreen/silead.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/input/touchscreen/silead.c b/drivers/input/touchscreen/silead.c
index 7fad43e..08d6549 100644
--- a/drivers/input/touchscreen/silead.c
+++ b/drivers/input/touchscreen/silead.c
@@ -47,6 +47,8 @@
 
 #define SILEAD_STATUS_OK	0x5A5A5A5A
 #define SILEAD_TS_DATA_LEN	44
+#define SILEAD_TS_READ_FIRST_DATA_LEN	12
+#define SILEAD_TS_THIRD_POINT_REGISTER	0x8c
 
 #define SILEAD_CLOCK		0x04
 #define SILEAD_CLOCK_OFF	0xB5
@@ -209,7 +211,14 @@ static void silead_ts_read_data(struct i2c_client *client)
 	int x, y, id, touch_nr = 0, ret, i, offset;
 	struct gsl_touch_info cinfo = { { 0 } };
 	ret = i2c_smbus_read_i2c_block_data(client, SILEAD_REG_DATA,
-					    SILEAD_TS_DATA_LEN, buf);
+					    SILEAD_TS_READ_FIRST_DATA_LEN, buf);
+	if (buf[0] > SILEAD_MAX_FINGERS)
+		buf[0] = SILEAD_MAX_FINGERS;
+	if (buf[0] > 2)
+		ret = i2c_smbus_read_i2c_block_data(client,
+			SILEAD_TS_THIRD_POINT_REGISTER,
+				(buf[0] - 2) * SILEAD_POINT_DATA_LEN,
+					buf + SILEAD_TS_READ_FIRST_DATA_LEN);
 	if (ret < 0) {
 		dev_err(dev, "Data read error %d\n", ret);
 		return;
-- 
1.9.1

